### PR TITLE
Deprecate nvidia/apex

### LINKED ISSUE #14416: Deprecate the nvidia/apex integration
## Proposed refactor

Deprecation:

- Deprecate the `ApexMixedPrecisionPlugin` and passing `Trainer(amp_backend=...)`. To be removed in 1.10
- Add deprecation notices to apex throughout our docs

Removal:
- Remove all the apex-related glue throughout the codebase
- Remove the apex installation from CI
- Remove apex from our docs

### Motivation

APEX AMP can be regarded as deprecated in favor of PyTorch AMP which Michael Carilli implemented and advocated in #1337.

Most developer activity in the nvidia/apex repository happen in either apex/transformer, apex/optimizers, tests/L0, and/or apex/contrib. apex/amp directory hasn't seen changes for about 2 years

Given the 2-year hibernation would make it almost impossible to resume the support for the different optimization levels to O2.

It's unclear whether any nvidia teams use our apex plugin internally.

And the nvidia team is unable to provide support for apex bugs.

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging PyTorch Lightning, Transformers, and Hydra.

cc @tchaton @rohitgr7 @carmocca @justusschock @awaelchli @akihironitta @kaushikb11 @borda

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes #14416

**PyTorch Lightning package changes**:
- Removes apex from the docs and the deprecated pieces below
- Updates tests to avoid the deprecation messages, or catch them
- Deprecates `NativeMixedPrecisionPlugin` in favor of `MixedPrecisionPlugin`
- Deprecates the `LightningModule.optimizer_step(..., using_native_amp=bool)` argument
- Deprecates `Trainer(amp_backend)`
- Deprecates `Trainer.amp_backend`
- Deprecates `Trainer(amp_level)`. This was already deprecated but the previous deprecation has been replaced with the one in this PR for consistency.
- Deprecates the `ApexMixedPrecisionPlugin`. The deprecation message triggered by Apex internally is replaced by ours.
- Deprecates the `AMPType` enum and updates the codebase to avoid it. The `_DeprecatedEnumMeta` class is re-introduced to do this.
- Deprecates `DeepSpeedPrecisionPlugin(amp_type, amp_level)` arguments.

**Lite package changes**:
- Renames the `NativeMixedPrecision` class to `MixedPrecision`
- Removes `DeepSpeedPrecision(amp_type, amp_level)` arguments and their support
- Remvoes the `AMPType` enum and its uses.

### Follow-up items (out of scope for this PR):
- Remove native_ from the appropriate filenames
- Remove "Native" from the FSDP class names

### Does your PR introduce any breaking changes? If yes, please list them.

None intended

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @tchaton @carmocca @justusschock @awaelchli

### CODE REVIEW

File: src/pytorch_lightning/plugins/precision/native_amp.py [hunk 1]
```diff
@@ -125,6 +124,17 @@ def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
             self.scaler.load_state_dict(state_dict)
 
 
+class NativeMixedPrecisionPlugin(MixedPrecisionPlugin):
```
Comment (carmocca) [inline src/pytorch_lightning/plugins/precision/native_amp.py:127]: Follow-up items (out of scope for this PR):

- Remove native_ from the filename
- Remove "native" from the FSDP classes

**Review by awaelchli — APPROVED** [review]

**Review by Borda — COMMENTED** [review]
:otter:

**Review by justusschock — APPROVED** [review]
